### ISSUE TITLE
Define standard library (a base context with functions)

### ISSUE THREAD

**Issue by jonasfj** [taskcluster/json-e#30]
@hammad13060,

I'm thinking that `json-e` should have some standard library... Well, a set of functions that a built-in.
What do you think?

My thinking is things like:
 * `sqrt(x)` for square root
 * `min/max(a,b)`
 * `str(x)` for converting to string
 * `uppercase(x)`
 * `lowercase(x)`
 * `abs(c)` absolute value
 * `hex(x)` to hexa decimal encoding
 * `floor/ceil(x)`
 * `len(x)` length of x
 * `sort(x)`
 * `pow(a, b)`
 * `parse(json)`
 * `stringify(json)`
 * `round(x)`
 * `sum(x)` compute sum of values in x
 * `codepoint(x)` get integer value of character `x`, where x is a string of length 1

I'm not sure if we system put them under a namespace like `builtin.pow` or `math.pow` or if we should name the otherwise, or which functions we might be missing...

But we probably need things like getting the length of a list, converting integer to string. Those we can't live without. I suspect a few others like uppercase/lowercase are useful too.

**Comment by jonasfj** [comment]
Note:
 * I have no idea how many functions we want built-in
 * Which functions is too much
 * Which functions we might be missing

Another option is to say that strings have some properties like: `"hello".length()`, `"hello".uppercase()`
`"hello".lowercase()` and similarly, numbers have some properties like `4.toString()` and arrays have `[1,2,3].length()`... That might be better as those names won't ever conflict with variable names in the `context`.

We could probably define things like:
```js
let builtinMethods = {}

builtinMethods.string = {}
builtinMethods.string.length = function() {
  return this.length;
}
```

Then inside our interpreter when we parse `expression '.' identifier` we do don't throw an error if the type isn't object... then we lookup in builtinMethods first, and if nothing is found there we throw an error.

**Comment by hammad13060** [comment]
I like the idea of having builtin helper methods 👍 
I'll rather go with the first approach because It is highly unlikely that someone will chose ```toLower``` or ```toUpper``` as property name. Along with that we will mention the built-in methods in the docs so that users of the library will choose the names accordingly. We can have methods in separate namespace if you want, lets say namespace be ```builtin``` so users only have to take care that none of the properties conflict with ```builtin```.

**Comment by jonasfj** [comment]
it depends how many built-in methods we add...
Also I suspect that people who uses json-e will inject some custom built-in methods into the context...

So it might be best to add built-ins as properties on values instead of built-in functions in the context.
If we add built-in as default functions in the context, then downstream context might overwrite us... Hmm... I guess that won't break compatibility, it'll just leave downstream in a weird state where their context partially overwrites a newer built-in function.
So it wouldn't be a breaking change.. but we could avoid risking that by adding them as methods on values. Which probably wouldn't be much harder.

Notably, I also see things like `5.toString()` as nice since is very close to javascript, C#, Java, etc... where as `str(...)` is super python specific.

But it also depends if we want to add a lot or a just a few built-in methods...

**Comment by hammad13060** [comment]
Hmm okay 👍 
but what about built in functions like min, max which cannot be expressed as properties of values ?

**Comment by jonasfj** [comment]
> but what about built in functions like min, max which cannot be expressed as properties of values ?

That's a good point, similarly we can't have `{key: 1}.keys()` because this `{keys: 1}.keys()` would be weird :)

Hmm... maybe the first question is how many built-in functions do we want?

 * upper/lower case for strings
 * toString for numbers
 * length of array and string
 * keys of a dictionary

I could also imagine:
 * array.map and array.filter, if we add support for function expressions (so maybe later)

Anything else that is critical?

**Comment by hammad13060** [comment]
I propose few [here](https://gist.github.com/hammad13060/1379f995178e7f3967599c5b492ea2ac) to start with. Feel free to propose more :)

**Comment by jonasfj** [comment]
Added a few comments here:
https://gist.github.com/jonasfj/68fa49a3f8a1e6a86d02995a98822827

Hmm, we probably want an option to do `[1,3,4].contains(3)` or maybe we should implement the `in` operator from python... Like `3 in [1,3,4,6]`... I'll file separately.

**Comment by jonasfj** [comment]
For the record I shift between either:
A) wanting to implement a huge standard library to allow as much power as possible,
B) wanting to keep the standard library as small as possible, and suggest that people write their own and inject it via the `context`.

Hmm, what to do?

**Comment by hammad13060** [comment]
I will  choose ```A```
  - things may get tricky on our side and we may pollute the context, but then user can confidently use functions like ```min```, ```max```, ```toString```, ```toLower``` etc, without worrying about their implementation and users can still implement their own functions by injecting them in context.

I think if we can go for ```B``` then going for ```A``` isn't a long leap forward

I'd like to know @djmitche thoughts on this as well.

**Comment by jonasfj** [comment]
Good point, let's get some input from @imbstack and @djmitche.

**Comment by djmitche** [comment]
I think we should start with a minimal product, and put it to use.  We will likely get a much more authentic picture of what features we really need at that point.  So I go for "B".

**Comment by jonasfj** [comment]
If we inject the standard library using `context = _.defaults({}, context,
builtinFunctions)`, then `context` entries will overwrite
`builtinFunctions`, and if we were to add additional builtin functions in
the future that wouldn't cause compatibility issues.
So I think that's a fair argument.

--
Regards Jonas Finnemann Jensen.

2017-02-21 9:51 GMT-08:00 Dustin J. Mitchell <notifications@github.com>:

**Comment by hammad13060** [comment]
Idea of generating context in this ```context = _.defaults({}, context, builtinFunctions)``` manner looks nice. Okay lets add few builtin's and start using the library :)

**Comment by djmitche** [comment]
It looks like this is implemented now, in #37.
